File: OC/内容层动画/WGOCCoreAnimation/WGOCCoreAnimation/CAGradientLayer动画/跳动动画/WGBeatAniViewController.m
//
//  WGBeatAniViewController.m
//  WGOCCoreAnimation
//
//  Created by wanggang on 2019/5/27.
//  Copyright © 2019 bozhong. All rights reserved.
//

#import "WGBeatAniViewController.h"

@interface WGBeatAniViewController ()

@property (nonatomic, strong) NSArray *colorArr;
@property (nonatomic, strong) NSMutableArray *layerArr;
@property (nonatomic, strong) NSTimer *timer;

@end

@implementation WGBeatAniViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = @"跳动动画";
    self.view.backgroundColor = [UIColor blackColor];
    
    self.timer = [NSTimer timerWithTimeInterval:0.5 target:self selector:@selector(colorChage) userInfo:nil repeats:YES];
    [[NSRunLoop currentRunLoop] addTimer:self.timer forMode:NSDefaultRunLoopMode];
}

- (void)colorChage{
    
    for (CAGradientLayer *layer in self.layerArr) {
        
        NSInteger index = arc4random_uniform(11);
        UIColor *color = self.colorArr[index];
        NSArray *arr = @[(id)UIColor.clearColor.CGColor, (id)color.CGColor];
        
        layer.colors = arr;
        layer.locations = @[@(0.0), @(1)];
        
        CABasicAnimation *ani = [[CABasicAnimation alloc] init];
        ani.keyPath = @"locations";
        CGFloat beginValue = arc4random_uniform(11)/10.0;
        ani.fromValue = @[@(beginValue), @(beginValue)];
        ani.toValue = @[@(1), @(1)];
        ani.duration = 0.5;
        [layer addAnimation:ani forKey:nil];
    }
}

- (NSArray *)colorArr{
    if (!_colorArr) {
        UIColor *color01 = UIColor.blackColor;
        UIColor *color02 = UIColor.blueColor;
        UIColor *color03 = UIColor.brownColor;
        UIColor *color04 = UIColor.cyanColor;
        UIColor *color05 = UIColor.grayColor;
        UIColor *color06 = UIColor.greenColor;
        UIColor *color07 = UIColor.magentaColor;
        UIColor *color08 = UIColor.orangeColor;
        UIColor *color09 = UIColor.purpleColor;
        UIColor *color010 = UIColor.redColor;
        UIColor *color011 = UIColor.yellowColor;
        _colorArr = @[color01, color02, color03, color04, color05, color06, color07, color08, color09, color010, color011];
    }
    return _colorArr;
}

- (NSMutableArray *)layerArr{
    if (!_layerArr) {
        
        _layerArr = [[NSMutableArray alloc] init];

        for (NSInteger i = 0; i< 15; i++) {
            
            CAGradientLayer *layer = [[CAGradientLayer alloc] init];
            
            CGFloat w = (SCREEN_WIDTH-10)/15;
            CGFloat h = 150;
            
            CGFloat x = 20;
            CGFloat y = 150;
            
            UIView *view = [[UIView alloc] initWithFrame:CGRectMake(w*i+x, y, w-x, h)];
            [self.view addSubview:view];
            
            layer.frame = view.bounds;
            layer.startPoint = CGPointMake(0, 0);
            layer.endPoint = CGPointMake(0, 1);
            
            [view.layer addSublayer:layer];
            [_layerArr addObject:layer];
        }
    }
    return _layerArr;
}


@end

File: Swift/内容层动画/WGCoreAnimation/WGCoreAnimation/CAGradientLayer动画/跳动动画/WGBeatAniViewController.swift
//
//  WGBeatAniViewController.swift
//  WGCoreAnimation
//
//  Created by wanggang on 2019/5/6.
//  Copyright © 2019 wanggang. All rights reserved.
//

import UIKit

class WGBeatAniViewController: UIViewController {

    var audioBarNum:Int = 0
    var gradientLayer:CAGradientLayer = CAGradientLayer()
    var layerArray:NSMutableArray = NSMutableArray()
    var colorArray:NSMutableArray = NSMutableArray()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        view.backgroundColor = UIColor.black
        title = "跳动动画"
        
        audioBarNum = 15

        buildColorArr()
        buildLayerArr()
        
        Timer.scheduledTimer(timeInterval: 0.5, target: self, selector: #selector(changeColor), userInfo: nil, repeats: true)
    }
    
    @objc func changeColor() {
        
        for layer in layerArray {
            
            let index = Int(arc4random_uniform(11))
            let color = colorArray[index] as! UIColor
            let colorArr = [UIColor.clear.cgColor, color.cgColor]
            
            let layer = layer as! CAGradientLayer
            layer.colors = colorArr
            layer.locations = [0.0, 1.0]
            
            let ani = CABasicAnimation()
            ani.keyPath = "locations"
            let beginValue = Float(arc4random_uniform(11))/10.0
            ani.fromValue = [beginValue, beginValue]
            ani.toValue = [1.0, 1.0]
            ani.duration = 0.4
            layer.add(ani, forKey: nil)
        }
    }
    
    func buildLayerArr() {
        
        for index in 0...audioBarNum {
            
            let h:CGFloat = 150
            let w:CGFloat = (self.view.frame.size.width-10)/CGFloat(audioBarNum)
            let x:CGFloat = 20
            let y:CGFloat = 150
            let view:UIView = UIView(frame: CGRect(x: w*CGFloat(index)+x, y: y, width: w-x, height: h))
            self.view.addSubview(view)
            
            gradientLayer = CAGradientLayer()
            gradientLayer.frame = view.bounds;
            gradientLayer.startPoint = CGPoint(x: 0, y: 0);
            gradientLayer.endPoint = CGPoint(x: 0, y: 1);
            view.layer.addSublayer(gradientLayer)
            layerArray.add(gradientLayer)
        }
    }
    
    func buildColorArr() {
        
        let  color1:UIColor = UIColor(red: 255.0 / 255.0, green: 127.0 / 255.0, blue: 79.0 / 255.0, alpha: 1.0)
        let  color2:UIColor = UIColor(red: 138.0 / 255.0, green: 206.0 / 255.0, blue: 245.0 / 255.0, alpha: 1.0)
        let  color3:UIColor = UIColor(red: 216.0 / 255.0, green: 114.0 / 255.0, blue: 213.0 / 255.0, alpha: 1.0)
        let  color4:UIColor = UIColor(red: 51.0 / 255.0, green: 207.0 / 255.0, blue: 48.0 / 255.0, alpha: 1.0)
        let  color5:UIColor = UIColor(red: 102.0 / 255.0, green: 150.0 / 255.0, blue: 232.0 / 255.0, alpha: 1.0)
        let  color6:UIColor = UIColor(red: 255.0 / 255.0, green: 105.0 / 255.0, blue: 177.0 / 255.0, alpha: 1.0)
        let  color7:UIColor = UIColor(red: 187.0 / 255.0, green: 56.0 / 255.0, blue: 201.0 / 255.0, alpha: 1.0)
        let  color8:UIColor = UIColor(red: 255.0 / 255.0, green: 163.0 / 255.0, blue: 0.0 / 255.0, alpha: 1.0)
        let  color9:UIColor = UIColor(red: 203.0 / 255.0, green: 93.0 / 255.0, blue: 92.0 / 255.0, alpha: 1.0)
        let  color10:UIColor = UIColor(red: 61.0 / 255.0, green: 226.0 / 255.0, blue: 210.0 / 255.0, alpha: 1.0)
        let  color11:UIColor = UIColor(red: 25.0 / 255.0, green: 146.0 / 255.0, blue: 255.0 / 255.0, alpha: 1.0)
        colorArray.add(color1)
        colorArray.add(color2)
        colorArray.add(color3)
        colorArray.add(color4)
        colorArray.add(color5)
        colorArray.add(color6)
        colorArray.add(color7)
        colorArray.add(color8)
        colorArray.add(color9)
        colorArray.add(color10)
        colorArray.add(color11)
    }
}
